port: Port to Minecraft 1.21.11

ShadowCat117 opened the pull request:
Masks (Circle for minimap) Can be saved for after 1.21.11 release

ShadowCat117 committed:
port: Change version to 1.21.11

ShadowCat117 committed:
port: Port Command packet changes (sendUnsignedCommand -> sendUnattendedCommand)

ShadowCat117 committed:
port: Port ConnectionMixin (PacketSendListener -> ChannelFutureListener)

ShadowCat117 committed:
port: Port LevelRendererMixins & beacon rendering (Mixin and rendering changes)

ShadowCat117 committed:
port: Port SoundManagerMixin (Return value added to method)

ShadowCat117 committed:
port: Port ArmSwingEvent drop (Player -> LivingEntity)

ShadowCat117 committed:
port: Port Item DataComponent and GameProfile changes)

ShadowCat117 committed:
port: Port FixCrosshairPositionFeature (rendering changes)

ShadowCat117 committed:
port: Port HoverEvent and ClickEvent changes

ShadowCat117 committed:
port: Port MobEffects (MOVEMENT_SLOWDOWN -> SLOWNESS)

ShadowCat117 committed:
port: Port SharedConstants (method name change)

ShadowCat117 committed:
port: Port various inventory/equipment access methods

ShadowCat117 committed:
port: Port PlayerInteractEvent (getCommandSenderWorld -> level)

ShadowCat117 committed:
port: Port NpcDialogueFeature (Shift packet changed)

ShadowCat117 committed:
port: Port Container type changes (int -> short/byte, HashedStack)

ShadowCat117 committed:
port: Port ContainerSetContentEvent (Packet class converted to record)

ShadowCat117 committed:
port: Port RaycastUtils (Added inflationAmount parameter)

ShadowCat117 committed:
port: Port Component changes (Component serialization & command argument)

ShadowCat117 committed:
port: Port ChatTimestampFeature (Mixin changes)

ShadowCat117 committed:
port: PlayerRenderState -> AvatarRenderState

ShadowCat117 committed:
port: Port key and mouse changes (Keybind category, KeyEvent, MouseButtonEvent)

ShadowCat117 committed:
port: Port CustomLoadingScreenFeature (ReceivingLevelScreen merged into LevelLoadingScreen)

ShadowCat117 committed:
port: Port LocalPlayerMixin (startRiding extra parameter)

ShadowCat117 committed:
port: Port ClientPacketListenerMixin (ReceivingLevelScreen merged into LevelLoadingScreen, records and RespawnData packet)

ShadowCat117 committed:
port: Port CapeLayerMixin (Rendering changes)

ShadowCat117 committed:
port: Port WingsLayerMixin (Mixin changes)

magicus commented on common/src/main/java/com/wynntils/mc/mixin/ChatComponentMixin.java:
```diff
@@ -106,40 +112,46 @@ private double screenToChatX(double x, Operation<Double> original) {
 
     @Inject(
-            method = "render(Lnet/minecraft/client/gui/GuiGraphics;IIIZ)V",
+            method = "method_71992", // 1st forEachLine lambda in render
```
That is not ideal. It seems like it can break with any MC update. Is there someway we can rewrite the mixin to hook it in a more stable entry point?

ShadowCat117 replied:
Just how we have to deal with lambdas in mixins afaik

magicus replied:
I was thinking that maybe we can inject our code where the lambda is called instead of in the lambda itself, like if the original code is like:
thisList.forEach(l -> doSomething(l))

we could inject around the entire forEach call, rather than the actual lambda. But maybe that does not work, or is too much trouble?

ShadowCat117 replied:
coehlrich pointed out that these could technically be changed to `lambda$render$0` and `lambda$render$1` which is a better way of doing it.

I still have issues with doing more complex mixins so I'll probably leave it as is, but obviously if you want to give it a go then feel free, I doubt I will be doing all of the work on this port anyway you're welcome to make any changes

magicus commented on common/src/main/java/com/wynntils/mc/mixin/ChatComponentMixin.java:
```diff
@@ -106,40 +112,46 @@ private double screenToChatX(double x, Operation<Double> original) {
 
     @Inject(
-            method = "render(Lnet/minecraft/client/gui/GuiGraphics;IIIZ)V",
+            method = "method_71991", // 2nd forEachLine lambda in render
```
Same here...

magicus commented on common/src/main/java/com/wynntils/mc/mixin/KeyMappingMixin.java:
```diff
@@ -1,31 +0,0 @@
-            at = @At("RETURN"))
-    private void initPost(String name, InputConstants.Type type, int i, String category, CallbackInfo ci) {
-        // This needs to go directly to KeyBindManager and not through Managers
-        KeyBindManager.initKeyMapping(category, CATEGORY_SORT_ORDER);
```
Did this move elsewhere, or how do we handle this now?

ShadowCat117 replied:
Since this doesn't compile yet I can't say for sure but it seems like when we register the category (https://github.com/Wynntils/Wynntils/pull/3567/commits/291c8e8b8636f6eebf1307404dafb9e3f72ea2ee#diff-f3f29995c5bd917f90c5e465b6a954fffb6386e68b7909f4b5a3cc1387dbc1c8R37) that does it for us

magicus commented on common/src/main/java/com/wynntils/mc/mixin/WingsLayerMixin.java:
```diff
@@ -21,20 +21,20 @@
+                    "submit(Lcom/mojang/blaze3d/vertex/PoseStack;Lnet/minecraft/client/renderer/SubmitNodeCollector;ILnet/minecraft/client/renderer/entity/state/HumanoidRenderState;FF)V",
             at = @At("HEAD"),
             cancellable = true)
     private void render(
```
Is vanilla renaming all their `render` methods to `submit`? If so, should we?

ShadowCat117 replied:
I've not looked into it too deeply so don't take this for certain but I believe the new system is something akin to they "submit" everything to be rendered and then somewhere along the pipeline it is all done. So when we do finish up porting the rendering side of things, we may end up using submit just by principal

magicus commented on common/src/main/java/com/wynntils/utils/wynn/ContainerUtils.java:
```diff
@@ -49,8 +50,16 @@ public static boolean openInventory(int slotNum) {
+        changedSlots.put(
+                clickedSlot,
+                HashedStack.create(
+                        new ItemStack(Items.AIR), McUtils.mc().getConnection().decoratedHashOpsGenenerator()));
```
Extract to a helper?

magicus commented on common/src/main/java/com/wynntils/features/chat/ChatTimestampFeature.java:
```diff
@@ -129,7 +130,7 @@ public void onnChatComponentRenderText(ChatComponentRenderEvent.Text event) {
                         extension.getTimestamp().get().a(),
                         0,
                         event.getRenderY(),
-                        16777215 + (event.getTextOpacity() << 24));
```
This just looks odd. Did they change how they interpret this argument?

ShadowCat117 replied:
It's been a while since I originally did the port of this specific mixin but I think it might just be that the original version didn't handle this as cleanly as it could have

magicus commented:
I looked through all changes. Most of them look like trivial refactorings on the part of vanilla. There were a few places where I raised my eyebrows, I've commented on them just to make sure we don't miss anything.

ShadowCat117 committed:
port: Port Cape & Elytra rendering (Rendering changes)

ShadowCat117 committed:
port: Port tooltip rendering mixins (renderTooltip -> setTooltipForNextFrame)

ShadowCat117 committed:
port: Port text rendering (Removed BufferedFontRenderer, use GuiGraphics)

May have missed a few places that still call a PoseStack version of the methods

ShadowCat117 committed:
port: Port Style (Font ResourceLocation -> FontDescription)

ShadowCat117 committed:
port: ScoreboardHandler (getScoreboard() method removed)

ShadowCat117 committed:
port: Port GameProfile access (Converted to Record)

ShadowCat117 committed:
port: Port ActivityModel (getSharedSpawnPos -> RespawnData)

ShadowCat117 committed:
port: Port AddEntityEvent (getXa, getYa, getZa -> getMovement Vec3)

ShadowCat117 committed:
port: Port DynamicTexture (Label Supplier parameter)

ShadowCat117 committed:
port: Port WynntilsAccountService (getMinecraftSessionService() -> services().sessionService())

ShadowCat117 committed:
port: Port screen key and mouse input changes (KeyEvent, MouseButtonEvent, CharacterEvent, InputWithModifiers)

ShadowCat117 committed:
port: Port WynntilsScreen (renderBackground override no longer needed)

ShadowCat117 committed:
port: Port ChatScreen opening (isDraft parameter)

ShadowCat117 committed:
port: Port Player Skin access (texture -> body.texturePath)

ShadowCat117 committed:
port: Port KeyboardUtils (getWindow -> handle)

ShadowCat117 committed:
port: Port LootrunCompiler (ToFloatFunction -> BoundedFloatFunction)

ShadowCat117 committed:
wip: Port minimap circle mask

ShadowCat117 commented:
As of the `chore: Make it compile` commit this now compiles so it will be easier to check ports are actually correct. All that commit does it comment out code that does not work yet, or replaces it with something to allow it to compile.

Some commented out sections in that commit belong in previous commits and were missed, some were commenting out sections from previous commits that did not work and need to be redone

ShadowCat117 committed:
fix: Fix invisible item count

ShadowCat117 committed:
fix: Fix NeoForge production environment check

ShadowCat117 committed:
port: Port Lootrun path rendering

ShadowCat117 committed:
port: Port PlayerRenderEvent (Mixin changes)

ShadowCat117 committed:
port: Port Range Visualizer circle rendering

ShadowCat117 committed:
fix: Render clear background on overlay screens

ShadowCat117 committed:
port: Port majority of RenderUtils (Removed BufferedRenderUtils)

ShadowCat117 committed:
port: Port screen tooltip rendering

ShadowCat117 committed:
port: Port Bulk Buy item render (no new GuiGraphics needed)

ShadowCat117 committed:
port: Port ContainerLabelRenderEvent (Mixin target return type changed)

ShadowCat117 committed:
port: Port/remove PoseStack translation/scales

ShadowCat117 committed:
port: Port ChatScreenCreateEvent (Move event posting to ChatComponentMixin, draft parameter)

ShadowCat117 committed:
port: Port Inventory ContainerRenderEvent

ShadowCat117 committed:
port: Port RenderSystem.recordRenderCall (Use McUtils.mc().execute)

ShadowCat117 committed:
port: Port ItemStackLayerRenderStateMixin (Method name and parameter changes)

ShadowCat117 committed:
port: Port PlayerRenderLayerEvent armour layers (Mixin changes, Avatar -> Humanoid render state)

ShadowCat117 commented:
Fruma has been delayed to early 2026 so we have a bit more time to work on this, it is also very likely we will be moving to 1.21.11 instead as that is expected to release this year.

As I mentioned on Discord I'm currently working on porting the custom nametag rendering, but after that the remaining rendering ports have given me a bit of trouble and I will likely need someone else to take a look.

And as for the remaining non-rendering ports/bugs, I've not looked into them too deeply so they may or may not be simple tasks.

ShadowCat117 committed:
fix: Fix hint message foreground shadow

ShadowCat117 committed:
port: Port map rendering

ShadowCat117 committed:
port: Port shaman totem tracking and highlights (Mixin and inventory changes)

ShadowCat117 committed:
port: Port cancelling entity rendering (Mixin changes)

ShadowCat117 committed:
port: Port translucent player rendering (Mixin changes)

ShadowCat117 committed:
fix: Add keybind category localisation

ShadowCat117 committed:
port: Port nametag rendering (Rendering and mixin changes)

ShadowCat117 committed:
port: Port leaderboard badge rendering

ShadowCat117 committed:
port: Port tooltip fitting (Rework events)

ShadowCat117 committed:
port: Port item compare (Change renderTooltip call, prevent tooltip fitting interruption)

ShadowCat117 committed:
port: Port ForgeLevelRendererMixin (Entity -> EntityRenderState)

ShadowCat117 commented:
From a brief look 1.21.11 didn't seem too bad so will start working on that now so this won't compile for a little bit in case anyone was trying it out

ShadowCat117 committed:
port: ResourceLocation -> Identifier

ShadowCat117 committed:
port: Port package changes

ShadowCat117 committed:
port: Port WynntilsButton (renderWidget -> renderContents)

ShadowCat117 committed:
port: Port Camera rotation/position/entity methods (Record alternatives)

ShadowCat117 committed:
port: Port ClientCommandSourceStack (Permissions level int -> PermissionSet)

ShadowCat117 committed:
port: Port screen init change (Removed Minecraft parameter)

ShadowCat117 committed:
port: Port WynncraftButtonFeature server ping (EventLoopGroupHolder parameter added)

ShadowCat117 committed:
port: Port AbstractContainerScreen renderSlot mixin (Mouse pos added)

ShadowCat117 committed:
chore: Disable DevAuth

Needs updating

ShadowCat117 committed:
chore: Remove lootrun path fabulous warning

ShadowCat117 committed:
port: Port item screenshots

ShadowCat117 committed:
port: Port map lootrun paths

ShadowCat117 commented:
@magicus As you have been working on the improved font handling do you want to take a crack at properly porting them? At the moment the `fontStyleInvalidFonts` test is failing with this, but even the ones that pass should probably be reworked:
```
Expected :§c§{f:minecraft:banner/nosuchfont}inherited font
Actual   :§c§{f:Resource[id=minecraft:banner/nosuchfont]}inherited font
```

Not sure how you want to handle the 3 font types (Resource, AtlasSprite and PlayerSprite) so don't want to step on your toes and change anything drastically as the font handling is still a bit wip

ShadowCat117 committed:
fix: Change default share item keybind to F4

ShadowCat117 committed:
refactor: Rework keybinds to fix them not saving

magicus commented:
Sorry, been out of the loop for a bit here. Sure, I can take a look at the font handling, I probably need to read up a bit on how Minecraft vanilla intends this to work. I am likely for forget it though, so feel free to remind me if you don't see anything happening here soon. :)

What is the timeline for us releasing the 1.21.11 port?

ShadowCat117 commented:
> What is the timeline for us releasing the 1.21.11 port?

Alongside the Wynncraft 2.2 (Fruma) update is the plan, it's currently targeting early 2026 so we've still got a bit of a time for the final issues. There might even be another game drop between now and then so we'll just go to whatever the latest is at that point

ShadowCat117 committed:
refactor: Add ItemUtils#createHashedItem

ShadowCat117 commented:
This is done now aside from the 3 remaining issues so I will stop the force pushes and any future fixes will just be comitted as normal. So if you want to do a re-review @magicus then by all means do so. Would recommended starting from the first commit again as those earlier ones you reviewed have likely changed. 

I'll mainly be focusing on cleaning it up and fixing any more issues I find whilst you work on the font handling and hopefully bolyai will have the time and knowledge to fix the other 2 issues I've not been able to port

ShadowCat117 committed:
fix: Fix Wynntils menu tag rendering

ShadowCat117 committed:
fix: Render backgrounds of player interaction buttons

ShadowCat117 committed:
fix: Fix guide import/export button size

ShadowCat117 committed:
fix: Add back pose translation to LootchestTextFeature

ShadowCat117 committed:
fix: Fix armor durability function crash

ShadowCat117 committed:
fix: Fix territory management screen crash

ShadowCat117 committed:
fix: Don't use mouse position for guild overall production widget position

ShadowCat117 committed:
fix: Fix only one tooltip rendering on item sharing and territory management screen

ShadowCat117 committed:
fix: Fix guild log button background not rendering

ShadowCat117 committed:
fix: Fix seaskipper destination info rendering

ShadowCat117 committed:
fix: Fix item lock and inventory emerald count rendering above compare tooltips

ShadowCat117 committed:
fix: Fix EventThread.IO message filtering

ShadowCat117 committed:
chore: Remove buffersource from overlays

ShadowCat117 committed:
fix: Fix reload button size

ShadowCat117 committed:
Revert "fix: Fix EventThread.IO message filtering"

This reverts commit ae3106198185d993df3039e3caa9d12b43d51f8a.

ShadowCat117 committed:
Reapply "fix: Fix EventThread.IO message filtering"

This reverts commit 13b043a91d5c830b644e00326ba3436b9d5b7dc8.

ShadowCat117 committed:
feat: Handle cursor in screens

ShadowCat117 committed:
fix: Restore float rendering

int was used originally when porting rendering, but was reverted to float later, these are just some leftovers

ShadowCat117 committed:
chore: Update mod menu version

ShadowCat117 committed:
chore: Bump Fabric & NeoForge

ShadowCat117 committed:
refactor: Rework AddCommandExpansionFeature & FilterAdminCommandsFeature to fix ClickEvents

ShadowCat117 committed:
fix: Fix missed float rendering case

ShadowCat117 committed:
fix: Fix colour rendering issues

ShadowCat117 committed:
Merge branch 'main' into 1.21.10

ShadowCat117 committed:
fix: Merge fixes

ShadowCat117 committed:
fix: Fix StoreTierAnnotator font check

ShadowCat117 committed:
fix: Don't render on LevelLoadingScreen

ShadowCat117 committed:
chore: Update parchment, bump fabric and neoforge

ShadowCat117 committed:
chore: Update DevAuth

ShadowCat117 committed:
feat: Restore original keybind behaviour

Keybinds belonging to disabled features are removed from the mappings

ShadowCat117 committed:
Merge branch 'main' into 1.21.10

ShadowCat117 committed:
fix: Merge fixes

ShadowCat117 committed:
fix: Don't run disabled keybinds

ShadowCat117 committed:
chore: Bump ModMenu

ShadowCat117 committed:
fix: Rework font handling

ShadowCat117 commented:
I have pushed something that works for the font handling, it may not be exactly how you want it to be, but just in case of the off chance we need to release sooner than expected. It is now just the circle minimap mask left to be ported which we may have to drop for now

ShadowCat117 committed:
Merge branch 'main' into 1.21.10

ShadowCat117 committed:
fix: Merge fix

ShadowCat117 committed:
fix: Use ; instead of : for separating FontDescription info

ShadowCat117 committed:
Merge branch 'main' into 1.21.10

ShadowCat117 committed:
fix: Merge fixes

ShadowCat117 committed:
chore: Disable circle minimap

ShadowCat117 committed:
fix: Check for null entity

Fixes https://github.com/Wynntils/Wynntils/issues/3674

ShadowCat117 committed:
fix: Fix armour hiding affecting armour stands

ShadowCat117 committed:
chore: Change share item default to F5

ShadowCat117 committed:
chore: Bump dependencies

ShadowCat117 committed:
Merge branch 'main' into 1.21.10

ShadowCat117 committed:
chore: Merge fixes

kristofbolyai commented:
> Copilot reviewed 299 out of 458 changed files in this pull request and generated no comments.

Okay :)

ShadowCat117 committed:
fix: Remove text shadow from guide aspect rarity

ShadowCat117 committed:
ci: spotless formatting

ShadowCat117 committed:
Merge branch 'main' into 1.21.10

ShadowCat117 committed:
chore: Change beta version

ShadowCat117 committed:
chore: Revert build.gradle version change

ShadowCat117 committed:
Merge branch 'main' into 1.21.10

ShadowCat117 committed:
chore: Bump NeoForge and EventBus version

magicus commented on common/src/main/java/com/wynntils/core/keybinds/KeyBindDefinition.java:
```diff
@@ -0,0 +1,501 @@
+    }
+
+    // region Chat
+    public static final KeyBindDefinition BOMB_RELAY_PARTY = register(
```
Hm, we're losing the distributed nature of our keybinds here. :-( It's not a super big deal, but I think it is a bit sad. Now the keybind definition is kind of split up between the default key and name here, and the function to call in the actual component implementing the functionality.

Do you think we could do something about it? Or is the late loading of components the problem that you needed to solve here?

magicus replied:
(And also, the keybind name/descriptions looks like they have hard-coded descriptions, like "Relay Bomb to Party". That seems like it should be subject to i18n treatment).

ShadowCat117 replied:
The issue was that we need keybinds to be loaded earlier, otherwise when the options file is loaded they get cleared so when we eventually load them they would be reset to their default key.

And yes, the point of me adding the id here is so that we can eventually convert these to using i18n as for whatever reason that wasn't done when the first keybinds were made. The issue is we need to upfix the options file as they are currently stored there using the plain English name, otherwise we'll cause a one time keybind reset

magicus replied:
And there's no way to keep the keybind info in the components themselves, but register with this new functionality? Or are the component classes (Features etc) loaded too late?

Good thinking about the i18n.

ShadowCat117 replied:
FeatureManager initialises them too late. We'd have to have a different method of locating them

magicus replied:
Ok, fair enough. We'll have to live with this.

magicus commented on common/src/main/java/com/wynntils/core/text/PartStyle.java:
```diff
@@ -177,13 +181,35 @@ public String asString(PartStyle previousStyle, StyleType type) {
+                                .append("{fr:")
+                                .append(fontCode)
+                                .append("}");
+                    } else if (font instanceof FontDescription.AtlasSprite(Identifier atlasId, Identifier spriteId)) {
```
Oooh, pattern matching! Me like. :)

magicus commented on common/src/main/java/com/wynntils/mc/mixin/ChatComponentMixin.java:
```diff
@@ -45,88 +40,95 @@ private void onInit(CallbackInfo ci) {
-        MixinHelper.post(new ChatComponentRenderEvent.Background(
-                (ChatComponent) (Object) this, guiGraphics, renderX, lineHeight, opacity));
+    @WrapOperation(
+            method = "method_75802",
```
I guess this is a lambda? It would be good to have some description of what this means in terms of the actual code in a comment.

ShadowCat117 replied:
Oops yeah I was trying to add those wherever it wasn't a standard name, missed this one

magicus commented on common/src/main/java/com/wynntils/mc/mixin/LevelRendererMixin.java:
```diff
@@ -54,50 +79,96 @@ private void renderLevelPost(
+                this.levelRenderState.cameraRenderState));
+    }
+
+    @Redirect(
```
`@Redirect` is not very friendly to other mods, and risks collisions. Is there some way to do this without it?

ShadowCat117 replied:
Maybe? I'll make a note to revisit it when I get a chance

magicus commented on common/src/main/java/com/wynntils/services/lootrunpaths/LootrunPathsService.java:
```diff
@@ -189,11 +188,6 @@ public void tryLoadLootrun(String fileName) {
             McUtils.sendMessageToClient(Component.translatable(
                             "service.wynntils.lootrunPaths.lootrunStart", start.getX(), start.getY(), start.getZ())
                     .withStyle(ChatFormatting.GREEN));
-
```
1) We should remove the i18n tag `service.wynntils.lootrunPaths.fabulousWarning` as well.

2) That warning states: "Fabulous graphics detected - lootrun paths may not be visible on this setting. Switch to fancy or fast graphics to see them."

Are we sure we do not have problems at certain graphic settings? Do we know what caused us to display this warning in the first place? Or has it just been ported over from Legacy?

ShadowCat117 replied:
1. It was removed in the same commit https://github.com/Wynntils/Wynntils/pull/3567/changes/bd7a6f583eff1d13f2271572816f0edc12b47774
2. The paths rendered fine with all of the different option settings I tried so I think it was just one of those things that got fixed in the port.

magicus replied:
Hm, I tried searching the code base for "fabulousWarning" and got a hit... I'll have to double check that I had checked out the correct version.

Ah, right, this is for the beta. I was on main. Nevermind then.

magicus commented on common/src/main/java/com/wynntils/utils/mc/McUtils.java:
```diff
@@ -178,6 +180,11 @@ public static void sendChat(String message) {
 
     public static void openChatScreen(String keybindCommand) {
-        mc().openChatScreen(keybindCommand);
+        // TODO: Improve mixin to not require event posting here
```
Is this TODO being tracked?

ShadowCat117 replied:
Not officially, I can open a issue for it

magicus replied:
Thanks.
